Triple indexed syntax

miaoyinb opened the pull request:
ref. #21930 

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1154646/) on c3518fe wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21931/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21931/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 52df555830acf877b3d4637e8abed5e2d72a5295`

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 5fb5bc70cf1b8894b363e5678171ff90c10472bc
+Subproject commit e3d9b2aa4e11ddecdf4fd1bd4d4e13f575e8c432
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1154646/) on c3518fe wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

miaoyinb commented:
Just a draft for now. Need to wait the libmesh PR and at least add a test object.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1154683/) on 5ec3801 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

miaoyinb commented on framework/src/parser/Parser.C:
```diff
@@ -1848,6 +1896,53 @@ Parser::setDoubleIndexParameter(const std::string & full_name,
+  // NOTE: the substrings are _not_ of type T yet
+  std::vector<std::string> first_tokenized_vector;
+  std::vector<std::vector<std::string>> second_tokenized_vector;
+  MooseUtils::tokenize(buffer, first_tokenized_vector, 1, "|");
```
I am using "|" for now but this is definitely open for discussion.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1154704/) on 5ec3801 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21931/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1154717/) on 5ec3801 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>52df55</th>
      <th colspan=4>#21931 5ec380</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.15%</td>
      <td>84.12%</td>
      <td>-0.03%</td>
      <td>38.46%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>78890</td>
      <td>78914</td>
      <td>+24</td>
      <td>25</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14862</td>
      <td>14896</td>
      <td>+34</td>
      <td>40</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21931/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21931/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21931/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 38.46% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 5fb5bc70cf1b8894b363e5678171ff90c10472bc
+Subproject commit b80b03bd739302597f41830edcda869ef0d4830c
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1155962/) on a2796f3 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

miaoyinb commented:
@GiudGiud @loganharbour 

The libMesh PR has been merged (https://github.com/libMesh/libmesh/issues/3380). Please advise how we should proceed here. Thanks.

miaoyinb commented on test/src/userobjects/ReadTripleIndex.C:
```diff
@@ -0,0 +1,518 @@
+  if (_real_tri != reference_values)
+    mooseError("Error reading real_tri.");
+
+  // check unsigned int
+  if (_uint_tri.size() != reference_values.size())
+    mooseError("Error reading uint_tri.");
+  for (unsigned int i = 0; i < _uint_tri.size(); i++)
+  {
+    if (_uint_tri[i].size() != reference_values[i].size())
+      mooseError("Error reading uint_tri.");
+    for (unsigned int j = 0; j < _uint_tri[i].size(); j++)
+    {
+      if (_uint_tri[i][j].size() != reference_values[i][j].size())
+        mooseError("Error reading uint_tri.");
+      for (unsigned int k = 0; k < _uint_tri[i][j].size(); k++)
+        if (_uint_tri[i][j][k] != static_cast<unsigned int>(reference_values[i][j][k] * 10.0))
+          mooseError("Error reading uint_tri.");
+    }
+  }
```
Will do, I was too lazy.

dschwen commented:
I'm not convinced this is a good idea. 3d arrays w will be large and unwieldy and not suitable for input parameters

miaoyinb commented:
> I'm not convinced this is a good idea. 3d arrays w will be large and unwieldy and not suitable for input parameters

We do have several cases that 3D arrays could make the input syntax clearer, including the element extra integer swap parameter in`FancyExtruderGenerator`.

GiudGiud commented:
The extrusion cases they have are indexed by axial plane, then hold both an origin and a target number. 
We re open to suggestions but this works imo

miaoyinb committed:
template routine ref #21930

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1156120/) on 77dcda9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21931/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21931/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format de5880a5b3225bb52505bf07c3e4b5be7542c5af`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1156120/) on 77dcda9 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

miaoyinb committed:
clang format fix ref #21930

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1156157/) on 3201152 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

YaqiWang commented:
> The extrusion cases they have are indexed by axial plane, then hold both an origin and a target number. We re open to suggestions but this works imo

Sometimes using map type for parameters can make the input more straightforward. If you have your own parsing function as I suggested in the issue, you can add whatever keywords to make the syntax more readable like
```
layer 1
x y z

layer 0
a b c
```
where `layer` can serve as a keyword. It is hard to be convincing that the 3-d vector type is needed.

miaoyinb commented:
> > The extrusion cases they have are indexed by axial plane, then hold both an origin and a target number. We re open to suggestions but this works imo
> 
> Sometimes using map type for parameters can make the input more straightforward. If you have your own parsing function as I suggested in the issue, you can add whatever keywords to make the syntax more readable like
> 
> ```
> layer 1
> x y z
> 
> layer 0
> a b c
> ```
> 
> where `layer` can serve as a keyword. It is hard to be convincing that the 3-d vector type is needed.

Actually,  in `FancyExtruderGenerator`, we need the triple-indexed vector form for the `elem_integers_swaps` parameter.
The idea originate from the subdomain swap idea. For example. if we have subdomain ids 1, 2, and 3 (e.g., fuel, insulation, reflector) in the original 2D mesh and want to extrude it into three layers of elevation. After extrusion, we would like to make all top and bottom layer to be subdomain 3 (reflector). To do so, we can set 

```
subdomain_swaps = '1 3 2 3 3 3;
                   1 1 2 2 3 3;
                   1 3 2 3 3 3'
```
or the simple form without unchanged pairs.

```
subdomain_swaps = '1 3 2 3 ;
                   ;
                   1 3 2 3'
```

Here `subdomain_swaps` is a 2D vector, the first dimension is extruded layer (i.e., elevation). the second dimension is pairs of subdomain ids that need to be swapped (the first is the original, the second is the wanted value).

To use the similar approach for element extra integers such as `material_id` and `deletion_id`. The major challenge is that we may have more than one set of element extra integers that need to be swapped. Currently, the sets that need to be swapped are defined by `elem_integer_names_to_swap`; and then the swap rules of multiple sets of integers are stacked in `elem_integers_swaps`. For example,

```
    elem_integer_names_to_swap = 'element_extra_integer_1 element_extra_integer_2'
    elem_integers_swaps = '1 4 2 8;
                           2 7;
                           1 6;      #first three lines are for element_extra_integer_1 
                           1 8 2 4;
                           2 5;
                           1 6'      #last three lines are for element_extra_integer_2
```
We think using a 3D vector instead of stacking them into a 2D vector would be helpful. Like

```
    elem_integer_names_to_swap = 'element_extra_integer_1 element_extra_integer_2'
    elem_integers_swaps = '1 4 2 8;
                           2 7;
                           1 6|      # "|" is used here for the new dimension
                           1 8 2 4;
                           2 5;
                           1 6'    
```

dschwen commented:
Three dimensional vectors are going to be unintuitive. Both one and two D can easily be laid out in the input. The third dimension is not easy to lay out intuitively in a 2D text file.

miaoyinb commented:
> Three dimensional vectors are going to be unintuitive. Both one and two D can easily be laid out in the input. The third dimension is not easy to lay out intuitively in a 2D text file.

This is true as we are all limited by 2D screens. 

Here we just want to have the third type of numeric separators (aside from space and semicolon) to make some input parameters clearer.

GiudGiud reviewed (commented):
3D-like (eg currently flattened in one axis) vectors are not new in MOOSE parameters and are less clear than true 3D vector input. 
The extrusion code has long had:
- 2D plane data, with an outer 1D axial indexing
- swapped pairs, with an outer 1D axial indexing

custom map parsing would on the contrary be completely new. Long term we need to think:
- what is easiest to QA?
- what is easiest to parse automatically for plugin-laguage-server assisted inputs?

YaqiWang commented:
> 

The input can be made into
```
    elem_integers_swaps = '
                          element_extra_integer_1
                           1 4 2 8;
                           2 7;
                           1 6

                           element_extra_integer_2
                           1 8 2 4;
                           2 5;
                           1 6' 
```
for example, with a custom parsing function. It can be even made into something like
```
    elem_integers_swaps = '
                          element_extra_integer_1
                          layer 0 1=>4 2=>8
                          layer 1 2=>7
                          layer 2 1=>6

                           element_extra_integer_2
                           layer 0 1=>8 2=>4
                           layer 1 2=>5
                           layer 2 1=>6' 
```

I do not have a super strong opinion on 3D vectors in inputs but I agree with @dschwen that they are very counter intuitive.

GiudGiud commented:
We could include the layer indexing or the extra_integer name as a comment and achieve the same. These two examples you are showing are 3D vectors, just with explicit declaration of the indexing on one or two indexes.

This custom parsing function is wishful thinking at this point. It would introduce as many difficulties as it would solve. i dont see it working well at all with the language-server work. 

3D is not **counter** intuitive. It's maybe not intuitive but it definitely does not defy or counter logic. It's in direct continuation of systems that are already in MOOSE.

YaqiWang commented:
> We could include the layer indexing or the extra_integer name as a comment and achieve the same. These two examples you are showing are 3D vectors, just with explicit declaration of the indexing on one or two indexes.
> 
> This custom parsing function is wishful thinking at this point. It would introduce as many difficulties as it would solve. i dont see it working well at all with the language-server work.
> 
> 3D is not **counter** intuitive. It's maybe not intuitive but it definitely does not defy or counter logic. It's in direct continuation of systems that are already in MOOSE.

The reason you were giving is not user oriented. As I said, if you insist, I am ok.

miaoyinb committed:
globalparam test and comment update ref #21930

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1157600/) on 051f7aa wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

GiudGiud commented on test/tests/parser/parse_triple_index/parse_triple_index.i:
```diff
@@ -0,0 +1,83 @@
+[]
+
+[Outputs]
+  exodus = true
```
no need for that, in fact no need for a problem solve either I think

GiudGiud reviewed (changes requested):
the parsing seems to be totally fine with double symbols ;; and || even though the dimension should be off??

;;;;;;;;;; is parsed incorrectly too. This kind of short syntax could be used by users who want to skip a plane for example in the extrusion

miaoyinb committed:
enable special triple index cases with empty vector ref #21930

miaoyinb commented:
> the parsing seems to be totally fine with double symbols ;; and || even though the dimension should be off??
> 
> ;;;;;;;;;; is parsed incorrectly too. This kind of short syntax could be used by users who want to skip a plane for example in the extrusion

Good point. I think I fixed this in the commit I just pushed and added some test cases.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1162793/) on 5425766 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1162916/) on 173245f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21931/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21931/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fa664410cea12f16d74333e1a608071309ee9ef2`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1162916/) on 173245f wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1162953/) on e2f5e39 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

miaoyinb committed:
Enable Triple Indexed Input Syntax

closes #21930

miaoyinb commented:
@GiudGiud Finally, conda has been updated so that this one can pass all the checks.

GiudGiud reviewed (approved):
looks great!

Thank you for the contribution and addressing the reviews promptly and extensively!
